File: app/src/main/java/rx_playground/com/jablonski/rxandroidplayground/modeldetails/ModelDetailsContract.java
package rx_playground.com.jablonski.rxandroidplayground.modeldetails;

import android.support.v4.app.Fragment;

import rx_playground.com.jablonski.rxandroidplayground.model.ImageSource;
import rx_playground.com.jablonski.rxandroidplayground.model.Model;
import rx_playground.com.jablonski.rxandroidplayground.model.Photo;
import rx_playground.com.jablonski.rxandroidplayground.model.RowConfig;

/**
 * Created by yabol on 15.05.2017.
 */

public class ModelDetailsContract {
    public interface Provider{
        RowConfig getObject(int position);
        int getCount();
        int getPositionType(int position);
    }

    public interface ImagesProvider{
        ImageSource getImage(int position);
        int getPhotosCount();
    }

    public interface Presenter{
        void loadModelData(String modelId);
        void displayModel(Model model);
        void displayPhotos(Photo photos);
        Model getModel();
    }

    public interface View{
        void displayModelDetails();
        void displayPhotos();
    }

    public interface Repository{
        void getModelDetails(String modelId);
        void getModelPhotos(String modelId);
    }
}

File: app/src/main/java/rx_playground/com/jablonski/rxandroidplayground/modeldetails/ModelDetailsContract.kt
package rx_playground.com.jablonski.rxandroidplayground.modeldetails

import io.reactivex.Observable
import rx_playground.com.jablonski.rxandroidplayground.model.*

/**
 * Created by yabol on 15.05.2017.
 */

class ModelDetailsContract {
    interface Provider {
        val count: Int
        fun getObject(position: Int): RowConfig
        fun getPositionType(position: Int): Int
    }

    interface ImagesProvider {
        val photosCount: Int
        fun getImage(position: Int): ImageSource?
    }

    interface Presenter {
        fun loadModelData(modelId: String)
        fun displayModel(model: Model)
        fun displayPhotos(photos: Photo?)
    }

    interface View {
        fun displayModelDetails(model: Model)
        fun displayPhotos()
    }

    interface Repository {
        fun getModelDetails(modelId: String): Observable<ModelDetailsResult>
        fun getModelPhotos(modelId: String): Observable<PhotosResult>
    }
}
